otherlibs/stdune/src/io.ml

```diff
@@ -254,32 +254,91 @@ struct
        it fails for w/e reason *)
     external copyfile : string -> string -> unit = "stdune_copyfile"
 
-    external available : unit -> bool = "stdune_is_darwin"
+    external sendfile : Unix.file_descr -> Unix.file_descr -> int -> unit
+      = "stdune_sendfile"
+
+    external darwin : unit -> bool = "stdune_is_darwin"
+
+    external linux : unit -> bool = "stdune_is_linux"
+
+    let available =
+      if darwin () then `Copyfile else if linux () then `Sendfile else `Nothing
+
+    let sendfile =
+      let setup_copy ?(chmod = Fun.id) ~src ~dst () =
+        match Unix.openfile (Path.to_string src) [ O_RDONLY ] 0 with
+        | exception Unix.Unix_error (Unix.ENOENT, _, _) -> Error `Src_missing
+        | ic ->
+          let open Result.O in
+          let+ size, oc =
+            let dst = Path.to_string dst in
+            try
+              let stat = Unix.fstat ic in
+              let perm = chmod stat.st_perm in
+              Ok
+                ( stat.st_size
+                , Unix.openfile dst [ O_WRONLY; O_CREAT; O_TRUNC ] perm )
+            with exn -> (
+              Unix.close ic;
+              match exn with
+              | Unix.Unix_error (Unix.EISDIR, _, _) -> Error `Dst_is_a_dir
+              | _ -> Error (`Exn (Exn_with_backtrace.capture exn)))
+          in
+          (ic, oc, size)
+      in
+      fun ?chmod ~src ~dst () ->
+        (* All of this exception translation is done for now to maintain the
+           same error messages as the other file copying functions.
+
+           Eventually, we should stop using exceptions for signalling these
+           errors. But that's a bit of a large change since there's a lot of
+           exception catching to audit. *)
+        match setup_copy ?chmod ~src ~dst () with
+        | Error (`Exn exn) -> Exn_with_backtrace.reraise exn
+        | Error `Dst_is_a_dir ->
+          let message =
+            Printf.sprintf "%s: Is a directory" (Path.to_string dst)
+          in
+          raise (Sys_error message)
+        | Error `Src_missing ->
+          let message =
+            Printf.sprintf "error: %s: No such file or directory"
+              (Path.to_string src)
+          in
+          raise (Sys_error message)
+        | Ok (ic, oc, size) ->
+          Exn.protectx (ic, oc, size)
+            ~f:(fun (i, o, size) -> sendfile o i size)
```

Comment:
re: readability (https://github.com/ocaml/dune/pull/7257/files?w=1#r1131466187) it felt weird to get `i,o,size` and send `o i size`.